### PR TITLE
refactor: rearrange remote API implementation classes to allow for different implementations for each API version

### PR DESCRIPTION
**Pull request by MohamadJaara**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

kalium does not support having multiple implementation for remote api interfaces

### Solutions

Rearrange remote API implementation classes to allow for different implementations for each API version

in the network module now v0 API implementation are separated from the abstraction layer this will allow to easily create new implementations for newer api versions.
 e.g. 

```kotlin

// abstraction layer exposed to logic
interface LogoutApi {
    suspend fun logout(): NetworkResponse<Unit>
    suspend fun removeCookiesByIds(removeCookiesByIdsRequest: RemoveCookiesByIdsRequest): NetworkResponse<Unit>
    suspend fun removeCookiesByLabels(removeCookiesByLabelsRequest: RemoveCookiesByLabels): NetworkResponse<Unit>
}

// v0 implementation
internal open class LogoutV0 internal constructor(
    private val authenticatedNetworkClient: AuthenticatedNetworkClient,
    private val sessionManager: SessionManager
) : LogoutApi {

    private val httpClient get() = authenticatedNetworkClient.httpClient

    override suspend fun logout(): NetworkResponse<Unit> = wrapKaliumResponse {
        // do stuff
    }

    override suspend fun removeCookiesByIds(removeCookiesByIdsRequest: RemoveCookiesByIdsRequest): NetworkResponse<Unit> =
        wrapKaliumResponse {
            // do stuff
        }

    override suspend fun removeCookiesByLabels(removeCookiesByLabelsRequest: RemoveCookiesByLabels): NetworkResponse<Unit> =
        wrapKaliumResponse {
            // do stuff
        }

}

// and v1 only needs to override the changed end points
internal open class LogoutV1 internal constructor(
    private val authenticatedNetworkClient: AuthenticatedNetworkClient,
    private val sessionManager: SessionManager
) : LogoutApiVo {

    override suspend fun removeCookiesByIds(removeCookiesByIdsRequest: RemoveCookiesByIdsRequest): NetworkResponse<Unit> =
        wrapKaliumResponse {
            // do different stuff than v0
        }

}

```

The next step will be to add v2 and implement all the changed endpoints
Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/auth/login/SSOLoginRepository.kt [hunk 1]
```diff
@@ -3,9 +3,9 @@ package com.wire.kalium.logic.data.auth.login
 import com.wire.kalium.logic.NetworkFailure
 import com.wire.kalium.logic.functional.Either
 import com.wire.kalium.logic.wrapApiRequest
-import com.wire.kalium.network.api.AuthenticationResultDTO
-import com.wire.kalium.network.api.user.login.SSOLoginApi
-import com.wire.kalium.network.api.user.login.SSOSettingsResponse
+import com.wire.kalium.network.api.base.model.AuthenticationResultDTO
+import com.wire.kalium.network.api.base.unAuthenticated.SSOLogin
```
Comment (typfel) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/auth/login/SSOLoginRepository.kt:7]:
```suggestion
import com.wire.kalium.network.api.base.unauthenticated.SSOLogin
```
Noticed that the casing wrong in a few places.

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/base/unAuthenticated/SSOLogin.kt [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package com.wire.kalium.network.api.base.unAuthenticated
+
+import com.wire.kalium.network.api.base.model.AuthenticationResultDTO
+import com.wire.kalium.network.utils.NetworkResponse
+
+interface SSOLogin {
```
Comment (typfel) [inline network/src/commonMain/kotlin/com/wire/kalium/network/api/base/unAuthenticated/SSOLogin.kt:6]: This interface is missing the `Api` suffix.

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/v0/authenticated/KeyPackageApiV0.kt [hunk 3]
```diff
@@ -1,14 +1,19 @@
...
+import com.wire.kalium.network.api.base.authenticated.keypackage.KeyPackageList
 import com.wire.kalium.network.utils.NetworkResponse
 import com.wire.kalium.network.utils.wrapKaliumResponse
 import io.ktor.client.request.get
 import io.ktor.client.request.parameter
 import io.ktor.client.request.post
 import io.ktor.client.request.setBody
 
-class KeyPackageApiImpl internal constructor(private val authenticatedNetworkClient: AuthenticatedNetworkClient) : KeyPackageApi {
+internal open class KeyPackageApiV0 internal constructor(private val authenticatedNetworkClient: AuthenticatedNetworkClient) : KeyPackageApi {
```
Comment (typfel) [inline network/src/commonMain/kotlin/com/wire/kalium/network/api/v0/authenticated/KeyPackageApiV0.kt:16]: We can later (after V2 introduced) move all MLS related APIs to V2, they shouldn't be accessible on V1 or V0.

**Review by typfel — CHANGES_REQUESTED** [review]
Great work! Just some minor comments.

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/v0/authenticated/KeyPackageApiV0.kt (hunk 3 above)
Comment (MohamadJaara) [reply network/src/commonMain/kotlin/com/wire/kalium/network/api/v0/authenticated/KeyPackageApiV0.kt:16]: I will add a TODO for now and disable it when v2 is implemented

**Review by typfel — APPROVED** [review]

**Pull request merged**